osp/osp.c

```diff
@@ -791,3 +1070,142 @@ osp_param_free (osp_param_t *param)
+ * @param[in]   port      CA public key.
+ * @param[in]   username  Credential username.
+ * @param[in]   password  Credential password.
+ *
+ * @return New osp credential.
+ */
+osp_credential_t *
+osp_credential_new (const char *type, const char *service, const char *port,
+                    const char *username, const char *password)
+{
+  osp_credential_t *new_credential;
+
+  new_credential = g_malloc0 (sizeof (osp_credential_t));
+
+  new_credential->type = type ? g_strdup (type) : NULL;
+  new_credential->service = service ? g_strdup (service) : NULL;
+  new_credential->port = port ? g_strdup (port) : NULL;
+  new_credential->username = username ? g_strdup (username) : NULL;
+  new_credential->password = password ? g_strdup (password) : NULL;
+
+  return new_credential;
+}
+
+/**
+ * @brief Free an OSP credential.
+ *
+ * @param[in]   credential  The credential to free.
+ */
+void
+osp_credential_free (osp_credential_t *credential)
+{
+  if (!credential)
+    return;
+
+  g_free (credential->type);
+  g_free (credential->service);
+  g_free (credential->port);
+  g_free (credential->username);
+  g_free (credential->password);
+  g_free (credential);
+}
+
+
+osp_target_t *
+osp_target_new (const char *hosts, const char *ports)
+{
+  osp_target_t *new_target;
+  new_target = g_malloc0 (sizeof (osp_target_t));
+
+  new_target->hosts = hosts ? g_strdup (hosts) : NULL;
+  new_target->ports = ports ? g_strdup (ports) : NULL;
+
+  return new_target;
+}
+
+void
+osp_target_free (osp_target_t *target)
+{
+  if (!target)
+    return;
+
+  g_slist_free_full (target->credentials,
+                     (GDestroyNotify) osp_credential_free);
+  g_free (target->hosts);
+  g_free (target->ports);
+  g_free (target);
+}
+
+void
+osp_target_add_credential(osp_target_t *target, osp_credential_t *credential)
+{
+  if (!target || !credential)
+    return;
+
+  target->credentials = g_slist_prepend (target->credentials, credential);
+}
+
+osp_vt_group_t *
+osp_vt_group_new (const char *filter)
+{
+  osp_vt_group_t *new_vt_group;
+  new_vt_group = g_malloc0 (sizeof (osp_vt_group_t));
+
+  new_vt_group->filter = filter ? g_strdup (filter) : NULL;
+
+  return new_vt_group;
+}
+
+void
+osp_vt_group_free (osp_vt_group_t *vt_group)
+{
+  if (!vt_group)
+    return;
+
+  g_free (vt_group->filter);
+  g_free (vt_group);
+}
+
+osp_vt_single_t *
+osp_vt_single_new (const char *vt_id)
+{
+  osp_vt_single_t *new_vt_single;
+  new_vt_single = g_malloc0 (sizeof (osp_vt_single_t));
+
+  new_vt_single->vt_id = vt_id ? g_strdup (vt_id) : NULL;
+  new_vt_single->vt_values = g_hash_table_new_full (g_str_hash, g_str_equal,
+                                                    g_free, g_free);
+
+  return new_vt_single;
+}
+
+void
+osp_vt_single_free (osp_vt_single_t *vt_single)
+{
+  if (!vt_single)
+    return;
+
+  g_hash_table_destroy (vt_single->vt_values);
+
+  g_free (vt_single->vt_id);
+  g_free (vt_single);
+}
+
+void
+osp_vt_single_add_value(osp_vt_single_t *vt_single,
```

Comment:
Missing space.

Reply:
Added the missing space in 6ad18f2060b0d9cb20d117605796a472a61179de